include/object_pool.h

```diff
@@ -0,0 +1,101 @@
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#ifndef DOSBOX_OBJECT_POOL_H
+#define DOSBOX_OBJECT_POOL_H
+
+#include "dosbox.h"
+
+// This class is a generic object re-use pool. Consider using it if your program
+// performs tens of thousands (or more) large object heap allocations as part of
+// its normal and ongoing runtime loop.
+//
+// Ongoing runtime loop:
+//     auto my_obj_ptr = new HugeType(arg1, arg2);
+//     .. do a bit of work ..
+//     delete my_obj_ptr;
+//
+// If you have this pattern, then the ObjectPool is a good candidate:
+//
+//  Instantiate a long-lived pool for your HugeType:
+//     ObjectPool<HugeType> pool = {};
+//
+//  Ongoing runtime loop:
+//     auto my_obj_ptr = pool.Acquire(arg1, arg2);
+//     .. do a bit of work ..
+//     pool.Release(my_obj_ptr);
+
+// Note: this is less about performance and more about fitting a well known
+// use-case (frequent and ongoing heap allocation and destruction of a large
+// object type) with a well known solution (object pool).  The before-and-after
+// benefit will be reduced call outs to the heap and less heap fragmentation.
+
+#include <cassert>
+#include <memory>
+
+// Clang 15 for macOS 12 and 13 is missing C++17's <memory_resource>, however
+// Clang 15 for macOS 14 has it. When the build system eventually uses macOS 14
+// for all Apple builds then the '#else' portions can be removed. In the
+// meantime, we simply fallback to new and delete for this defficient compiler.
+#ifdef HAVE_MEMORY_RESOURCE
+	#include <memory_resource>
+#endif
+
+template <typename T>
+class ObjectPool {
+public:
+	ObjectPool() = default;
+	~ObjectPool() = default;
+
+	// Acquire a new object: allocate and construct it given the arguments.
+	template <typename... Args>
+	constexpr T* Acquire(Args&&... args)
+	{
+#ifdef HAVE_MEMORY_RESOURCE
+		auto ptr = static_cast<T*>(pool.allocate(sizeof(T), alignof(T)));
+		AllocTraits::construct(allocator, ptr, std::forward<Args>(args)...);
+#else
+		auto ptr = new T(std::forward<Args>(args)...);
+#endif
+		assert(ptr);
+		return ptr;
+	}
+
+	// Release the object: destruct it and deallocate its used memory.
+	constexpr void Release(T* ptr)
+	{
+		assert(ptr);
+#ifdef HAVE_MEMORY_RESOURCE
+		AllocTraits::destroy(allocator, ptr);
+		pool.deallocate(ptr, sizeof(T), alignof(T));
+#else
+        delete ptr;
+#endif
+	}
+
+#ifdef HAVE_MEMORY_RESOURCE
+private:
+	std::pmr::unsynchronized_pool_resource pool = {};
+	std::pmr::polymorphic_allocator<T> allocator{&pool};
+	using AllocTraits = std::allocator_traits<std::pmr::polymorphic_allocator<T>>;
```

Comment:
I have no idea what you're doing here, but if it works, it works 😅 Looking forward to having macOS 14 as the baseline, then we just simply use the stdlib.

Reply:
I understand that this has something to do with an allocator but @interloper98 knows more about this than we do :laughing: 

I've never used these things either.

Reply:
The polymorphic memory stuff (the PMR namespace) is template heavy and quite rough on the eyes :laughing: ; but the pieces only fit together in certain ways (otherwise it fails to compile). 

Mostly it lets you customise how a type handles memory when it's allocated, constructed, destructed, and deallocated. The `allocator{&pool}` hooks up the pool resource to the allocator (which is where the re-use comes into play).

All memory 'resources' need to provide standard API hooks for the allocator to use it (otherwise it won't compile). And in our case, we're just using C++'s own pool resource, but I suppose some people write their own resource pools (who knows.. maybe allocating memory from storage or across networks or crazy stuff like that).

All of the above so far only covers allocating and de-allocate bulk memory from a pool for our template type - but to be object oriented, we also need to call the constructor and destructor routines on the re-used chunk of memory. 

With that, our Acquired object will have the old contents. This is where  the `AllocTraits` comes into play. It lets us run the template type's constructor and destructor against the reused block of memory (including passing constructor options to it).

That's about it! Thankfully all the hard work is done under the hood by the C++ libraries, and we're gluing the pieces together! :jigsaw: 

I'll add a couple references to the documentation for these in the code, just for interested readers.

Reply:
Thanks 😅 Formidable stuff 😆

Reply:
(Added some comments about how the pool and allocator interact, and pointed to the ref docs for more details)